src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java

```diff
@@ -112,14 +159,76 @@ public void createBuildPlanForExercise(ProgrammingExercise programmingExercise,
         final String projectName = programmingExercise.getProjectName();
         final boolean recordTestwiseCoverage = Boolean.TRUE.equals(programmingExercise.isTestwiseCoverageEnabled()) && "SOLUTION".equals(planKey);
 
-        Plan plan = createDefaultBuildPlan(planKey, planDescription, projectKey, projectName, repositoryUrl, testRepositoryUrl, programmingExercise.getCheckoutSolutionRepository(),
-                solutionRepositoryUrl, auxiliaryRepositories)
-                .stages(createBuildStage(programmingExercise.getProgrammingLanguage(), programmingExercise.getProjectType(), programmingExercise.getPackageName(),
-                        programmingExercise.hasSequentialTestRuns(), programmingExercise.isStaticCodeAnalysisEnabled(), programmingExercise.getCheckoutSolutionRepository(),
-                        recordTestwiseCoverage, programmingExercise.getAuxiliaryRepositoriesForBuildPlan()));
+        String assignedKey = null;
+
+        boolean couldCreateCustomBuildPlan = false;
+        if (aeolusBuildPlanService.isPresent() && programmingExercise.getBuildPlanConfiguration() != null) {
+            try {
+                assignedKey = tryToPublishCustomBuildPlan(projectKey + "-" + planKey, programmingExercise, planDescription, repositoryUrl, testRepositoryUrl, solutionRepositoryUrl,
+                        auxiliaryRepositories);
+                if (assignedKey != null) {
+                    couldCreateCustomBuildPlan = true;
+                }
+            }
+            catch (ContinuousIntegrationBuildPlanException e) {
+                LOGGER.error("Could not create custom build plan for exercise " + programmingExercise.getTitle() + " with id " + programmingExercise.getId()
+                        + ", will create default build plan", e);
+            }
+        }
+        if (!couldCreateCustomBuildPlan) {
+            Plan plan = createDefaultBuildPlan(planKey, planDescription, projectKey, projectName, repositoryUrl, testRepositoryUrl,
+                    programmingExercise.getCheckoutSolutionRepository(), solutionRepositoryUrl, auxiliaryRepositories)
+                    .stages(createBuildStage(programmingExercise.getProgrammingLanguage(), programmingExercise.getProjectType(), programmingExercise.getPackageName(),
+                            programmingExercise.hasSequentialTestRuns(), programmingExercise.isStaticCodeAnalysisEnabled(), programmingExercise.getCheckoutSolutionRepository(),
+                            recordTestwiseCoverage, programmingExercise.getAuxiliaryRepositoriesForBuildPlan()));
+            bambooServer.publish(plan);
+            assignedKey = plan.getKey().toString();
+        }
+
+        setBuildPlanPermissionsForExercise(programmingExercise, assignedKey);
+    }
```

Comment:
I'm not too happy with this mix of things that happen here in one method. Can you extract the Aeolus-based creation into a new method and simplify the logic in here? I think this would dramatically improve the readability.